decision-records/2024-10-09--adding-k8s-labels-from-libs.md

```diff
@@ -0,0 +1,68 @@
+# Adding k8s Labels From Libraries
+**Date:** 2024-07-16  
+**Authors:** @dstathis @ca-scribner @IbraAoad
+
+## Context and Problem Statement
+
+To mark a pod as belonging to a given mesh, we need to apply a few labels.
+For example, with istio we would need to e.g.:
+```python
+{
+  "istio.io/dataplane-mode": "ambient",
+  "istio.io/use-waypoint": waypoint_name,
+  "istio.io/use-waypoint-namespace": waypoint_k8s_namespace,
+}
+```
+
+In order for charms to join the mesh, the beacon charm sends this set of k8s labels over relation data that the charm must apply to
+its pods. The `service_mesh` library should attempt to manage those labels.
+
+Juju manages the labels of all deployments, so charm code must re-apply the labels after key lifecycle events:
+
+1. Labels must match relation data after a leader change
+2. Labels must match relation data after a pod restart
+3. Labels must match relation data when relation data changes
+4. Labels applied by anything outside the library should remain untouched
+
+The issue seems to be when 1 and 3 happen at the same time (or 2 and 3). We will have no idea how the labels have
+changed. For example, lets say relation data is `{"foo": "bar"}` and labels are set correctly. Then the relation data is
+set reset to `{}` but the the leader pod dies at the same time. The new leader will come up and process the relation
+changed event. It will see that it has no stored state and that matches relation data. So it will do nothing. The
+`{"foo": "bar"}` label is now left on the pods unmanaged.
+
+The issue stems from the fact that StoredState is scoped to the pod and not the charm. Even if `use_juju_for_storage` is
+set to True, this issue will still persist. The usual solution would be to use a peer relation instead of StoredState
+but since this is a library, we can't really impose this on the charm (or can we?)
+
+
+## Accepted Solution
+
+We should create a ConfigMap to be used as a way to store state. Thus we will have state which lives for the lifetime
+of the charm and does not rely on controller storage (`use_juju_for_storage=True`). It also asks nothing extra from the charm developer.
+
+The naming schema, to guarantee uniqeness, is: `juju-service-mesh-{self._charm.app.name}-labels`.
```

Comment:
Based on the discussion [here](https://github.com/canonical/observability/pull/211/files#r1844382503), if we assert that using the service-mesh library will always create/modify if they exist the given labels (`istio.io/dataplane-mode`, etc), then I suggest we use this solution:

```suggestion
Each charm unit should manage the labels on its own Pod, and keep track of what it has been managing using StoredState.  The library implementing this would follow the procedure:

1. get `_service_mesh_cache` from StoredState (schema shown below).  If it doesn't exist, default value is `{managed_labels: {}}`. 

StoredState: {
	_service_mesh_cache: {
		managed_labels: {
			label1: value1,  # We only need labels here.  Could omit the values
			...
		}
	}
}

2. get `suggested_labels` from service mesh relation
3. for each key in `_service_mesh_cache` that is not in `suggested_labels`, delete that label from the Pod and delete it from the cache
4. for each key in `suggested_labels`, put those labels on this charm's Pod and save them to `_service_mesh_cache`

This procedure should execute on the following events:

* `service-mesh-relation-changed` (so that we catch any new/changed labels): run the procedure as outlined above
* `start` (so that we are guaranteed to run after a Pod restart, as that does not create a relation-changed event): do the procedure above, except since Pod restarts wipe our labels **we know we haven't set any managed labels yet**.  This means we don't need to track what happened before the restart and can always reset `_service_mesh_cache={managed_labels: {}}`.  Doing this reset is important so we don't accidentally inherit some stale cache from a previous execution.
```

To me, this is a good compromise between complexity and robustness.  Its possible we clobber someone's labels, but its unlikely (are they seriously setting their own istio labels?) and we warned them.

If we do not want to make this assertion and instead avoid clobbering their labels, I think there's a solution that's very similar to the above but manages to respect the labels.  Its more complex, but I believe it exists.  I haven't wrote it out because it is more complicated and I think not worth the effort, but if we want that solution instead I can write it up